### PR TITLE
test: refactor tmp dir helper fixtures

### PR DESCRIPTION
**Pull request by Suor**
- `tmp_dir` descends from Path
- convenient template methods to generate files and dirs, add them to
dvc and git and commit to git
- independent of old basic_env structures
- can go in any order
- modular: start with empty `tmp_dir`, initialize git with `scm`, dvc
repo with `dvc` add "repo template" with `repo_template`, ...

### CODE REVIEW

File: tests/func/test_add.py [hunk 1]
```diff
@@ -474,23 +470,18 @@ def test(self):
...
-    assert len(stages) == 1
-
-    foo_stage_file = repo_dir.FOO + Stage.STAGE_FILE_SUFFIX
-
-    # corrupt stage file
-    repo_dir.create(foo_stage_file, "this will break yaml structure")
+def test_should_cleanup_after_failed_add(tmp_dir, scm, dvc, repo_template):
+    # Add and corrupt a stage file
+    dvc.add("foo")
+    tmp_dir.gen("foo.dvc", "- broken\nyaml")
```
[the remark this answers is not in the document]
Comment (Suor) [reply tests/func/test_add.py]: I prefer to be more obvious.

File: tests/func/test_api.py [hunk 2]
```diff
@@ -145,6 +147,7 @@ def _set_remote_url_and_commit(repo, remote_url):
     repo.scm.commit("modify remote")
 
 
+# FIXME: this test doesn't use scm ;D
```
[the remark this answers is not in the document]
Comment (Suor) [reply tests/func/test_api.py]: We will step on each others toes if you start fixing this) This is mere a note to myself.

**Comment by Suor** [comment]
@pared The idea is that helpers will make generating a file so easy that you won't require files precreated with fixtures, which distribute logic over code. In your foo example:
```python
def test_add_file(tmp_dir, dvc):
    stage, = temp_dir.dvc_gen("foo")
    assert os.path.exists(stage.relpath)
```
So `foo` fixture saves 0 lines.

File: tests/dir_helpers.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+    def scm_add(self, filenames, commit=None):
+        self._require("scm")
+        filenames = _coerce_filenames(filenames)
+
+        self.scm.add(filenames)
+        if commit:
+            self.scm.commit(commit)
+
+    # Introspection mehtods
```
Comment (efiop) [inline tests/dir_helpers.py]:
```suggestion
    # Introspection methods
```

**Comment by Suor** [comment]
... *(sorry for breaking this)*...

This way you may use specific telling names for files, e.g. not reuse `code.py` as some pregitadded file. Bootstrap specific file structure and play with it in a single place. This should make tests more self-contained and hence more readable.

Getting read of `repo_template` might be even better, but it looks like it makes harder to migrate. I also wanted to show it to you, guys. So this remains a design question, do we need to get rid of template?

**Comment by Suor** [comment]
@mroutis may you take a look?

File: tests/dir_helpers.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        return str(self)
+
+    def _require(self, name):
+        if not hasattr(self, name):
+            raise TypeError(
+                "Can't use {name} for this temporary dir. "
+                'Did you forget to use "{name}" fixture?'.format(name=name)
+            )
+
+    def gen(self, struct, text=""):
```
Comment (mroutis) [inline tests/dir_helpers.py]: really neat; hard to read at the beginning but the interface is awesome :heart_eyes: 
would you mind documenting the intended usage, @Suor? (a quick summary on top of the module is enough))

**Comment by mroutis** [comment]
@Suor , what's currently missing from the implementation? (asking because of WIP in title)

File: tests/dir_helpers.py (hunk 4 above)
Comment (Suor) [reply tests/dir_helpers.py]: This is a good idea.

**Comment by Suor** [comment]
@mroutis It could be merged like this. We will, however, get 3 types of tests this way: unittests, pytest old, pytest new. I thought migrating the rest of the pytes tests first. The thing is the longer this PR lingers here the harder it would be to rebase it. So maybe merging it as is and migrating the rest of the pytest-style tests in a subsequent PR would be easier. What do you think @efiop?

**Comment by efiop** [comment]
@Suor A bit hesitant because of those reasons, but it is indeed reasonable. Let's do that, but let's also put the converting on the next sprint, so that we don't leave the mess for long. Are you willing to give it shot? 🙂

**Comment by Suor** [comment]
@efiop I am in.

**Comment by efiop** [comment]
@Suor Sounds good :)

**Comment by Suor** [comment]
So, should we merge it? I can add a usage module doc-string in a separate PR.

**Comment by efiop** [comment]
@Suor At least need to remove WIP suffix, I'm mostly waiting for that.

**Comment by Suor** [comment]
Done;

**Comment by efiop** [comment]
@Suor Please create a ticket for the further steps and add it to our sprint, so we can keep track of it.

**Pull request merged**